File: lostmap.cpp
#include <algorithm>
#include <set>
#include <iostream>
#include <vector>
#include <string>

using namespace std;

struct edge {
    int u;
    int v; 
    int d;
};

bool cmp(edge& x, edge& y) {
    return x.d < y.d;
}

int find(vector<int>& parent, int i) {
    if (parent[i] == i) {
        return i;
    }
    return find(parent, parent[i]);
}

void uni (vector<int>& parent, vector<int>& rank, int x, int y){
    int xroot = find(parent, x);
    int yroot = find(parent, y);

    if (rank[xroot] < rank[yroot]){
        parent[xroot] = yroot;
    } else if (rank[xroot] > rank[yroot]) {
        parent[yroot] = xroot;
    } else {
        parent[yroot] = xroot;
        rank[xroot]++;
    }
}


int main() {
    int n;
    cin >> n;

    vector<edge> edges;
    for (int i=0; i < n; i++) {
        for (int j=0; j < n; j++) {
            int num;
            cin >> num;
            if (j > i){
                edges.push_back({i, j, num});
            }
        }
    }
    sort(edges.begin(), edges.end(), cmp);

    vector<int> parent;
    vector<int> rank;
    for (int i=0; i < n; i++) {
        parent.push_back(i);
        rank.push_back(0);
    }
    
    for (auto i : edges) {
        int x = find(parent, i.u);
        int y = find(parent, i.v);
        if (x != y) {
            cout << i.u + 1 << " " << i.v + 1 << endl;
            uni(parent, rank, x, y);
        }
    }
}

File: lostmap.py
import heapq


def find(parent, i):
    if parent[i] == i:
        return i
    return find(parent, parent[i])

def union(parent, rank, x, y):
    xroot = find(parent, x)
    yroot = find(parent, y)

    if rank[xroot] < rank[yroot]:
        parent[xroot] = yroot
    elif rank[xroot] > rank[yroot]:
        parent[yroot] = xroot
    else:
        parent[yroot] = xroot
        rank[xroot] += 1

def kruskal(edge_list, n):
    parent = []
    rank = []
    for i in range(n):
        parent.append(i)
        rank.append(0)
    edges = 0
    while edges < n-1: 
        d, i, j = edge_list.pop() 
        x = find(parent, i)
        y = find(parent, j)
        if x != y:
            edges += 1
            print(i+1, j+1)
            union(parent, rank, x, y)
            
    
n = int(input())
edge_list = []
for i in range(n):
    row = list(map(int, input().split()))
    for j in range(i+1, n):
        edge_list.append((row[j], i, j))

kruskal(sorted(edge_list, reverse=True), n)
